### ISSUE TITLE
convex_hull of polytopes does not remove redundant vertices

### ISSUE THREAD

**Issue by schillic** [JuliaReach/LazySets.jl#561]
Example:
```julia
julia> using LazySets, Polyhedra
julia> v1 = [1., 0.]; v2 = [0., 1.]; v3 = [-1., 0.]; v4 = [0., -1.]; v5 = [0., 0.];
julia> p1 = VPolytope([v1, v2, v5]);
julia> p2 = VPolytope([v3, v4]);
julia> ch = convex_hull(p1, p2);
julia> vl = vertices_list(ch)
5-element Array{Array{Float64,1},1}:
 [1.0, 0.0] 
 [0.0, 1.0] 
 [0.0, 0.0] 
 [-1.0, 0.0]
 [0.0, -1.0]
```
I would say that `v5` should not occur in the hull, at least not in its `vertices_list`.

**Comment by schillic** [comment]
On the other hand, even the constructor of `VPolytope` does not remove redundant vertices.

**Comment by mforets** [comment]
Seems to be fixed now, see related PR #944.

**Comment by schillic** [comment]
No, it is not fixed yet. It either needs to be fixed upstream or we need to merge that PR :wink:

**Comment by schillic** [comment]
There is also this:

> the constructor of VPolytope does not remove redundant vertices.

Here is a proposal for a default constructor, but we have to make sure that we pass `false` internally whenever we know that no redundant vertices exist.

```julia
    function VPolytope{N}(vertices::Vector{Vector{N}};
                          removeredundancy::Bool=true) where {N<:Real}
        if removeredundancy
            @assert isdefined(@__MODULE__, :Polyhedra) "redundancy removal " *
                "needs the package 'Polyhedra' loaded"
            P = polyhedron(Polyhedra.vrep(vertices))
            removevredundancy!(P)
            vertices = P.vrep.points.points
        end
        return new{N}(vertices)
    end
```

**Comment by mforets** [comment]
Ok, got it. I like the idea of making `VPolytope` to remove redundant vertices, which goes in line with the similar constructor for `VPolygon`.

The "drawback" is that `Polyhedra` needs to be loaded (and also `CDDLib` to make `removevredundancy!` fast for high dims) every time one creates `VPolytope`, though-- there are of course other uses of `VPolytope` without `Polyhedra`, such as computing support vectos--.  Anyway those are first-time-to-compile overhead and the kwarg to the constructor is useful to suppress that calls.

---

The lines inside the `removeredundancy` can be shortened if we add https://github.com/JuliaReach/LazySets.jl/issues/952.

For the name of the argument we use `apply_convex_hull` in `VPolygon`, we should use the same in both.

**Comment by schillic** [comment]
> The "drawback" is that Polyhedra needs to be loaded (and also CDDLib to make removevredundancy! fast for high dims) every time one creates VPolytope, though-- there are of course other uses of VPolytope without Polyhedra, such as computing support vectos

I see `VPolytope` as a wrapper around the `Polyhedra` type (for instance, we *do need* `Polyhedra` to compute the support vector :stuck_out_tongue_winking_eye:) rather than a standalone type. And as you mentioned, a user can still bypass that call.

> For the name of the argument we use apply_convex_hull in VPolygon, we should use the same in both.

Agreed. I am in favor of `remove_redundancy` because it describes what happens (and not what is called in the background).

**Comment by schillic** [comment]
The original bug was fixed in #944.
I moved the discussion about removing redundant vertices in the `VPolytope` constructor to #1392.
